text/0000-decorators.md

```diff
@@ -0,0 +1,1212 @@
+   some amount of repetition in the decorator name and field name, or a
+   redundant field name:
+
+    ```js
+    class Foo {
+      @actions actions = {
+        onClick() { /* ... */ }
+      }
+    }
+    ```
+
+2. Decorate the class itself with the actions object as a parameter. This would
+   be awkward, since actions would be removed from the class definition:
+
+    ```js
+    @actions({
+      onClick() { /* ... */ }
+    })
+    class Foo {}
+    ```
+
+3. Decorate the class itself, with the expectation that the `actions` class
+   field exists and is an object. This leads to a disconnect between the
+   decoration and the definition that is easy to miss, and could be
+   counterintuitive to newcomers:
+
+    ```js
+    @actionHandler
+    class Foo {
+      actions = {
+        onClick() { /* ... */ }
+      }
+    }
+    ```
+
+None of these options is ergonomic in the least. Instead, it is much cleaner and
+easier to decorate method definitions that are directly on the class body:
+
+```js
+class Foo {
+  @action
+  onClick() { /* ... */ }
+}
+```
+
+This would be implemented by creating the class's `actions` object and assigning
+the method to it, setting up inheritance and such in the process. The decorator
+leaves the method definition on the class, where it can be called like a normal
+method would, including `super` functionality. This maintains compatibility with
+the legacy object model, while making the `actions` namespace an
+_implementation_ detail rather than something users need to know about.
+
+This does mean that actions can once again collide with actual lifecycle and
+event hooks on the class, since they are no longer namespaced. The decorator
+_could_ remove the method from the class definition entirely, but this would be
+confusing for users not familiar with the old `actions` namespace - why does
+this method disappear from the class? It would also break `super` functionality,
+so it would not be ideal to do this.
+
+The `@action` decorator _could_ warn users when it collides with a lifecycle
+hook. However, hooks may vary from class type to class type, which presents a
+design challenge. We could either:
+
+A. Allow classes to specify lifecycle hooks, and throw whenever `@action`
+   collides with a specified hook.
+B. Only throw on hooks that are shared across all classes, such as `init` and
+   `destroy`.
+C. Do nothing, and leave it to user's to know which lifecycle hooks exist.
+
+Specifying hooks for each class would be time consuming and could fall out of
+sync with the implementations. Throwing on "universal" hooks only is
+inconsistent, and could lead users to think that they are safe when they are
+not. This RFC suggests that we choose option C for this reason. When the actions
+namespace was introduced, lifecycle hooks like `didUpdateAttrs` and
+`didInsertElement` were less commonly known and used (event listeners were not
+uncommon). Most Ember users know they exist now, and will be aware that
```

Comment:
These hooks were never really the gotcha. The two biggest gotchas were:

1. `destroy`, which is a very tempting action name (e.g. destroy a todo item in a list) but will always get invoked when the component is destroyed, much to people's surprised.
2. DOM event hooks like `click()`. It's very common to have an action called `click`, but if you have a component method with this name, it will automatically get called if there's a click _anywhere_ inside the component, not just where you attached the action.

In either case, I think we can safely enumerate this list and lint if you have an action with the same name.